module/zfs/spa.c

```diff
@@ -984,19 +976,56 @@ spa_taskqs_init(spa_t *spa, zio_type_t t, zio_taskq_type_t q)
 		value = MIN(zio_taskq_batch_pct, 100);
 		break;
 
+	case ZTI_MODE_SCALE:
+		flags |= TASKQ_THREADS_CPU_PCT;
+		/*
+		 * We want more taskqs to reduce lock contention, but we want
+		 * less for better request ordering and CPU utilization.
+		 */
+		cpus = MAX(1, boot_ncpus * zio_taskq_batch_pct / 100);
+		if (zio_taskq_batch_tpq > 0) {
+			count = MAX(1, (cpus + zio_taskq_batch_tpq / 2) /
+			    zio_taskq_batch_tpq);
+		} else {
+			/*
+			 * Prefer 6 threads per taskq, but no more taskqs
+			 * than threads in them on large systems.
+			 */
```

Comment:
> The following table looks considerably different, particularly for large cpu counts. It's the result of setting `zio_taskq_batch_tpq = 6` and using the `count` set in the `zio_taskq_batch_tpq > 0)` branch of the conditional. What's the advantage of performing this alternate calculation for `count`? 

It is not so much alternative.  It just does rounding to the nearest instead of instead of down, exactly following the requested tpq value.  You must be reading it wrong.

> ```
>         total           taskq   taskq
> cpus    threads taskqs  threads percent
> ------- ------- ------- ------- -------
> 128     63      21      3       3 %
> 256     86      43      2       1 %
> ```

How have you got such results?  For me it is:
128: cpus = 128*0.8 = 102; count = (102 + 3) / 6 = 17; value = 80/17 = 4%; that gives 17 taskqs of 5 threads, that is 85 threads.
256: cpus = 256*0.8 = 204; count = (204 + 3) / 6 = 34; value = 80/34 = 2%; that gives 34 taskqs of 5 threads, that is 170 threads.

But so big number of taskqs comparing to small threads per queue makes no sense, since it will be very difficult to make all of them busy.  I left this mode only for people who do really want to experiment and know what they are doing, in which case I am just trying to not interfere.

> As an alternative what do you think about making the hard coded value of `6` configurable with `zio_taskq_batch_tpq` instead in the `else` clause? This could be explained in the documentation as "The target number of worker threads per taskq...".
> 
> ```diff
> - 			count = 1 + cpus / 6;
> + 			count = 1 + cpus / zio_taskq_batch_tpq ;
> ```

It is not so much target in that case as you have shown, varying from 3 to 12 instead of 6.  To me it makes the tuning pretty pointless, since system still decides too much on its own.

Reply:
> ```
>         total           taskq   taskq
> cpus    threads taskqs  threads percent
> ------- ------- ------- ------- -------
> 64      48      8       6       10%
> 128     88      11      8       7%
> 256     192     16      12      5%
> ```

You've got math wrong here too.  Due to the loop there the number of taskqs should not be bigger than number of threads per taskq (at least due to rounding problems as in case of 256 CPUs).
```
        total           taskq   taskq
cpus    threads taskqs  threads percent
------- ------- ------- ------- -------
64      48      6       8       13%
128     100     10      10       8%
256     168     14      12      5%
```

Reply:
> at least if not rounding problems as in case of 256 CPUs

I've just added rounding "value" to the nearest percent instead of down.  It should slightly improve handling of very low percents (too large number of takqs).  I thought about it from the beginning, but haven't thought it would really go that far.

Reply:
I've added corrected table into the comments.

Reply:
Thanks for double checking this.  Upon revisiting this I found the bug in my math and was able to verify the table you included is correct.  For the curious reader I had overlooked that `cpus` is not the same as `boot_ncpus` here!

Reply:
I also had some trouble following this code. Thanks for the conversation.